imap_processing/cdf/tests/test_cdf_attribute_manager.py

```diff
@@ -1,58 +1,322 @@
+        test_get_global_attrs["Logical_source_description"]
+        == "IMAP Mission TEST one document Level-T1."
+    )
+    # Not given, and not required information
+    assert test_get_global_attrs["Data_level"] is None
+    with pytest.raises(KeyError):
+        assert test_get_global_attrs["bad_name"] == "false info"
+
+    # Testing second elif statement
+    test_error_elif = cdf_manager.get_global_attributes("imap_test_T3_test")
+    with pytest.raises(KeyError):
+        assert test_error_elif["Data_type"] == "Does Not Exist"
+
+    # Load in more data using get_global_attributes
+    test_get_global_attrs_2 = cdf_manager.get_global_attributes("imap_test_T2_test")
+    # Testing information previously loaded into global attributes (first if statement)
+    assert test_get_global_attrs_2["Project"] == "STP>Solar-Terrestrial Physics"
+    # Testing first elif statement
+    assert test_get_global_attrs_2["Descriptor"] == "TEST>Testinstrument"
+    # "Data_type" not required according to default schema
+    assert test_get_global_attrs_2["Data_type"] == "T2_test-two>Test-2 test two"
+
+    # Testing how instrument_id operates
+    assert test_get_global_attrs["Project"] == cdf_manager.global_attributes["Project"]
+    assert (
+        test_get_global_attrs["Source_name"]
+        == cdf_manager.global_attributes["Source_name"]
+    )
+    assert (
+        test_get_global_attrs["Data_type"]
+        == cdf_manager.global_attributes["imap_test_T1_test"]["Data_type"]
+    )
+    assert (
+        cdf_manager.global_attributes["imap_test_T1_test"]["Logical_source"]
+        == "imap_test_T1_test"
+    )
+    with pytest.raises(KeyError):
+        assert cdf_manager.global_attributes["imap_test_T1_test"]["Project"]
+
+    # Trying to update a default global using get_global_attributes does not work.
+    # For example, thing about DOI event.
+
+    # Testing that required schema keys are in get_global_attributes
+    for attr_name in cdf_manager.global_attribute_schema.keys():
+        required_schema = cdf_manager.global_attribute_schema[attr_name]["required"]
+        if required_schema is True:
+            assert attr_name in test_get_global_attrs.keys()
+
+
+def test_add_global_attribute():
+    # Initialize CdfAttributeManager object which loads in default info
+    cdf_manager = CdfAttributeManager(Path(__file__).parent.parent / "config")
+
+    # Change filepath to load test global attributes
+    cdf_manager.source_dir = Path(__file__).parent.parent / "tests"
+    cdf_manager.load_global_attributes("imap_default_global_test_cdf_attrs.yaml")
+    cdf_manager.load_global_attributes("imap_test_global.yaml")
+
+    # Changing a dynamic global variable
+    cdf_manager.add_global_attribute("Project", "Test Project")
+    assert cdf_manager.global_attributes["Project"] == "Test Project"
+    test_get_global_attrs = cdf_manager.get_global_attributes("imap_test_T1_test")
+    assert test_get_global_attrs["Project"] == "Test Project"
```

Comment:
Nice test, I like it. Can you also add a test for adding a new attribute that didn't exist in the file? For example, delete one of the required attributes (you can delete it out of cdf_manager.global_attributes) and then assert that `get_global_attributes` fails for that attribute, then run `add_global_attribute` and assert that it now exists.

Reply:
Yeah! You will see the updated test in some of my new pushes, but I did have a few questions about this:
- The code in get_global_attributes does not throw an error if a required attribute is missing, it just sets that attribute to `None`. Shall this be changed in the method definition?
- I can't add attributes that are instrument specific because all instrument specific data is accessed through dictionary objects. and not CdfAttributeManager objects (such as `cdf_manager.global_attributes["instrument"]` or `instrument = cdf_manager.get_global_attributes("instrument)`). Is that okay?

Reply:
Ah. I think I got it now, thanks!